File: example/modules/perf-lab/android/src/main/java/expo/modules/perflab/PerfLabModule.kt
package expo.modules.perflab

import android.os.Build
import android.util.Log
import android.view.Display
import com.facebook.react.bridge.ReactContext
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

/**
 * Native measurement module for the performance lab (performance/README.md).
 *
 * Frame recording starts and stops on the main thread because that is the
 * thread whose frame intervals are measured.
 */
class PerfLabModule : Module() {
  private var recorder: FrameRecorder? = null
  private var jsQueueProbe: JsQueueProbe? = null

  override fun definition() = ModuleDefinition {
    Name("PerfLab")

    AsyncFunction("startFrames") {
      recorder?.stop()
      recorder = FrameRecorder(appContext.currentActivity, ::currentDisplay).also { it.start() }
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("stopFrames") {
      val active = recorder
      recorder = null
      active?.stop() ?: FrameRecorder.emptyRecording(currentDisplay())
    }.runOnQueue(Queues.MAIN)

    AsyncFunction("memorySnapshot") {
      ProcessStats.memorySnapshot(requireContext())
    }

    AsyncFunction("processStats") {
      ProcessStats.processStats(requireContext())
    }

    AsyncFunction("deviceInfo") {
      ProcessStats.deviceInfo(requireContext(), appContext.currentActivity, currentDisplay())
    }.runOnQueue(Queues.MAIN)

    Function("nowNs") {
      System.nanoTime().toDouble()
    }

    AsyncFunction("startJsQueueProbe") { intervalMs: Double ->
      jsQueueProbe?.stop()
      val context = requireContext() as? ReactContext
        ?: throw IllegalStateException("JS queue probe needs a ReactContext")
      jsQueueProbe = JsQueueProbe(context, intervalMs.toLong().coerceAtLeast(1)).also { it.start() }
    }

    AsyncFunction("stopJsQueueProbe") {
      val active = jsQueueProbe
      jsQueueProbe = null
      active?.stop() ?: mapOf(
        "latenessMs" to DoubleArray(0),
        "samples" to 0,
        "intervalMs" to 0.0,
        "durationMs" to 0.0,
      )
    }

    // `adb shell am start -n <pkg>/.MainActivity --es perfRun <url>` is an
    // alternative to the VIEW intent that the CLI uses.
    Function("launchRequest") {
      appContext.currentActivity?.intent?.getStringExtra("perfRun")
    }

    Function("probesAvailable") {
      ProbeBridge.isAvailable()
    }

    AsyncFunction("setProbesEnabled") { enabled: Boolean ->
      ProbeBridge.setEnabled(enabled)
    }

    AsyncFunction("drainProbes") {
      ProbeBridge.drainJson()
    }

    AsyncFunction("logLine") { line: String ->
      Log.i(TAG, line)
    }

    AsyncFunction("writeResultFile") { name: String, content: String ->
      val context = requireContext()
      val base = context.getExternalFilesDir(null) ?: context.filesDir
      val directory = File(base, "perf-lab").also { it.mkdirs() }
      val file = File(directory, name)
      file.writeText(content)
      file.absolutePath
    }
  }

  private fun requireContext() =
    requireNotNull(appContext.reactContext) { "React context is not available" }

  private fun currentDisplay(): Display? {
    val activity = appContext.currentActivity ?: return null
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
      activity.display
    } else {
      @Suppress("DEPRECATION")
      activity.windowManager.defaultDisplay
    }
  }

  private companion object {
    const val TAG = "NitroMapsPerfLab"
  }
}

File: example/modules/perf-lab/ios/PerfLabModule.swift
import ExpoModulesCore
import Foundation
import UIKit
import os.log

/// Native measurement module for the performance lab (performance/README.md).
///
/// Frame recording starts and stops on the main thread because that is the
/// thread whose frame intervals are measured.
public final class PerfLabModule: Module {
  private static let log = OSLog(subsystem: "com.nitromaps.perflab", category: "results")
  private var recorder: FrameRecorder?

  public func definition() -> ModuleDefinition {
    Name("PerfLab")

    AsyncFunction("startFrames") { () -> Void in
      _ = self.recorder?.stop()
      let recorder = FrameRecorder()
      recorder.start()
      self.recorder = recorder
    }.runOnQueue(.main)

    AsyncFunction("stopFrames") { () -> [String: Any] in
      guard let recorder = self.recorder else {
        return FrameRecorder.emptyRecording()
      }
      self.recorder = nil
      return recorder.stop()
    }.runOnQueue(.main)

    AsyncFunction("memorySnapshot") { () -> [String: Any] in
      ProcessStats.memorySnapshot()
    }

    AsyncFunction("processStats") { () -> [String: Any] in
      ProcessStats.processStats()
    }.runOnQueue(.main)

    AsyncFunction("deviceInfo") { () -> [String: Any] in
      DeviceInfo.collect()
    }.runOnQueue(.main)

    Function("nowNs") { () -> Double in
      Double(DispatchTime.now().uptimeNanoseconds)
    }

    // `xcrun simctl launch <udid> <bundle> --perf-run=<url>` passes the run
    // request without the "Open in …?" prompt that `simctl openurl` shows.
    Function("launchRequest") { () -> String? in
      for argument in ProcessInfo.processInfo.arguments where argument.hasPrefix("--perf-run=") {
        return String(argument.dropFirst("--perf-run=".count))
      }
      return ProcessInfo.processInfo.environment["PERF_LAB_RUN"]
    }

    Function("probesAvailable") { () -> Bool in
      ProbeBridge.isAvailable
    }

    AsyncFunction("setProbesEnabled") { (enabled: Bool) -> Void in
      ProbeBridge.setEnabled(enabled)
    }

    AsyncFunction("drainProbes") { () -> String in
      ProbeBridge.drainJSON()
    }

    AsyncFunction("logLine") { (line: String) -> Void in
      os_log("%{public}@", log: Self.log, type: .default, line)
    }

    AsyncFunction("writeResultFile") { (name: String, content: String) -> String in
      let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
      let directory = documents.appendingPathComponent("perf-lab", isDirectory: true)
      try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
      let file = directory.appendingPathComponent(name)
      try content.write(to: file, atomically: true, encoding: .utf8)
      return file.path
    }
  }
}
